Split HUD from lib code

owalk opened the pull request:
Change from lib side singleton design pattern to lib side global object design pattern.
Allow for constructing hud definitions within the client using tools defined in the lib.

owalk committed:
Remove hud from lib side code
Allows for definition of client side HUD.

owalk committed:
Fix minotaur collision errors

Minotaur was missing collision definitions.

owalk committed:
Remove forgotten debug log

A proper logging system has not yet been created.

Sean-Nishi commented on m1wengine/abstract_HUD.py:
```diff
@@ -0,0 +1,150 @@
+
+    Class for holding all game score information.
+
+    Attributes
```
You have to add in docstrings for all the new vars in init

Sean-Nishi commented on m1wengine/abstract_HUD.py:
```diff
@@ -0,0 +1,150 @@
+    _boredom_meter: int
+        The current boredom meter for the game
+
+    Methods
```
Be sure to include return value in docstrings, even if it is None. Also add it in where the methods are implemented.

Sean-Nishi commented on m1wengine/abstract_HUD.py:
```diff
@@ -0,0 +1,150 @@
+    def display_level_hint(self, new_value: bool):
+        """Set the display level hint value.
+
+        Parameters
```
Need return value type hint and docstring here

Sean-Nishi commented on m1wengine/abstract_HUD.py:
```diff
@@ -0,0 +1,150 @@
+        return self._level_paused
+
+    @pause_level.setter
+    def pause_level(self, new_value: bool):
```
Need return value type hint and docstring here

owalk commented:
oops, this wasn't ready yet. Still self reviewing

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,163 @@
+
+    Class for holding all game score information.
+
+    Attributes
```
Update docstrings.

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,163 @@
+        Pause level value setter
+    """
+
+    SCORE_BOX_X_PX = 20
```
Type hints for all vars in this new file

Sean-Nishi commented on m1wengine/abstract_HUD.py:
```diff
@@ -0,0 +1,150 @@
+        self._show_level_hint: bool = False
+        self._level_paused: bool = False
+        self.level_prompt_timer: int = 0
+        self._level_box = pygame.sprite.Sprite()
+        self._level_hint_txt = pygame.sprite.Sprite()
```
typehints

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,163 @@
+        score_sprite_list.append(self.__current_score_sprite)
+
+        # define current boredom meter
+        self.__boredom_meter_spr: Text = Text(
```
No var mangling.

Sean-Nishi commented on m1wengine/abstract_HUD.py:
```diff
@@ -0,0 +1,150 @@
+            if not self._level_hint_added_flag:
+                self._level_hint_added_flag = True
+
+                level_hint_sprites = self._hud_builder.build_level_hint_box(
```
Magic numbers 500, 400. Also, Refactor build_level_hint_box to accept a tuple instead of two magic ints

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,163 @@
+
+        return score_sprite_list
+
+    def build_level_hint_box(
```
Refactor params to accept a tuple or x, y instead of two ints

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,163 @@
+        self, x_location: int, y_location: int, hint_str: str = "No string provided"
+    ) -> list[pygame.sprite.Sprite]:
+        """Create all sprites for the level hint box."""
+        level_hint_sprites = []
```
type hint

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,163 @@
+        self, x_location: int, y_location: int, method
+    ) -> list[pygame.sprite.Sprite]:
+        """Create pause button."""
+        pause_button_sprites = []
```
type hint

Sean-Nishi reviewed (changes requested):
Mostly type hints and docstring. I would look over every method just to be sure.

owalk committed:
Update level manager to Global Object Design

owalk committed:
Update NPC to use new HUD

owalk committed:
Add units to NPC vars

Sean-Nishi commented on m1wengine/abstract_HUD.py:
```diff
@@ -0,0 +1,154 @@
+    -------
+    __init__(self)
+        Initialize the singleton object's starting instance
+    start_prompt_timer(self)
```
One more return value

Sean-Nishi commented on m1wengine/abstract_HUD.py:
```diff
@@ -0,0 +1,154 @@
+        self._hud_builder: HudBuilder = HudBuilder()
+        self._show_level_hint: bool = False
+        self._level_paused: bool = False
+        self.level_prompt_timer: int = 0
```
Why does this var not have prepending "_"? other vars like _level_paused are referenced once in this file yet have the _. Be sure to update the docstrings if/when you make this change.

owalk replied:
good catch, I thought it was used in NPC and so public, but instead, it's done through the methods defined in abstracthud. client-side it's also only used in the children

This should be changed to a pretending underscore
![image](https://github.com/Sean-Nishi/M1Wengine/assets/10580211/b261a369-322c-481d-b7d9-5c3fba404d29)

owalk replied:
oh this was why...
![image](https://github.com/Sean-Nishi/M1Wengine/assets/10580211/e232a796-f606-4041-a132-02930001c11c)

From NPC. I'll add a method to get the value

owalk replied:
once again nevermind, the above code lives in abstract hud. this is good with an underscore

Sean-Nishi commented on m1wengine/abstract_HUD.py:
```diff
@@ -0,0 +1,154 @@
+
+    @abstractmethod
+    def start_prompt_timer(self):
+        """Start the timer for displaying level prompts."""
```
docstrings for "raises"

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,176 @@
+        """
+        SCORE_BOX_HEIGHT: int = 50
+        SCORE_BOX_WIDTH: int = 150
+        x = coordinates[0]
+        y = coordinates[1]
```
type hints

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,176 @@
+            The text to display within the level hint
+        """
+        level_hint_sprites: list = []
+        x = coordinates[0]
+        y = coordinates[1]
```
type hints

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,176 @@
+    ) -> list[pygame.sprite.Sprite]:
+        """Create pause button.
+
+        Parameters
```
Need "Returns" part of docstring

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,176 @@
+            A bound method to run on button clicks
+        """
+        pause_button_sprites: list = []
+        x = coordinates[0]
+        y = coordinates[1]
```
type hints

Sean-Nishi replied:
Actually, can reference these directly when creating the button on lines 152, 153. Do you want to make this reference change for no temp x, y vars?

owalk replied:
they can be referenced directly all over the place in hud builder within all 3 methods that use them. The var names are only there for code clarity so you know what data is stored at index 0 and index 1.

With direct references we would need at least some comments added in. This way is easier to read

owalk replied:
will add type hints

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,176 @@
+    def add_box(self, rect: pygame.rect.Rect) -> pygame.sprite.Sprite:
+        """Create a background box for HUD elements.
+
+        Parameters
```
Need "Returns" docstring

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,176 @@
+    ) -> list[pygame.sprite.Sprite]:
+        """Create all sprites for the level hint box.
+
+        Parameters
```
Returns docstring

Sean-Nishi commented on m1wengine/builders/hud_builder.py:
```diff
@@ -0,0 +1,176 @@
+    def build_score_info(self, coordinates: tuple) -> list[pygame.sprite.Sprite]:
+        """Create all sprites for the HUD score info.
+
+        Parameters
```
Returns docstring

Sean-Nishi reviewed (changes requested):
Lots of docstring and type hint stuff. Also when building the parts of the hud, can reference coordinates method param directly instead of creating a temp x, y var to hold those values

owalk committed:
The starting state should be patrol with the current logic

the NPC will otherwise not patrol until the state changes.
Radar methods may neeed to be adjusted to go back to this logic.
